### DISCUSSION TITLE
implementing xep-0313

### DISCUSSION THREAD

**Discussion by thegodtune** [xmppjs/xmpp.js#920] [Ideas]
Hi, @sonnyp here is the new discussion.
All help and suggestions are well appreciated. 
Thanks.

**Comment by sonnyp** [comment]
Hi,

Please expand, what is the use case? which platform?
Why do you think it belong in xmpp.js ? 

https://github.com/xmppjs/xmpp.js/issues/773

**Reply by thegodtune** [reply]
The use case is matrix-Bifrost, we are trying to implement [Message Archive Management](https://xmpp.org/extensions/xep-0313.html) on matrix-Bifrost and since xmpp.js is the backend being used I thought it will be right to check if the solution would be to extend xmpp.js.

**Reply by sonnyp** [reply]
Implement it in matrix-Bitfrost directly.

xmpp.js embraces xmpp/xml instead of building abstractions on top so I recommend users to abstract what they need rather than try and create an abstraction that fits all. If you want to build a generic MAM library for xmpp.js I'm sure that'd make a few users happy but it's out of scope for xmpp.js see https://github.com/xmppjs/xmpp.js/issues/531#issuecomment-528846223

I'll happily review it and include it in the README.md though.

xmpp.js goal is to provide high level primitives to build xmpp powered applications, here is a pointer on how to get started

```js
// See client/example.js first

// if you can/want to use jsx
const result = await client.iqCaller.request(
  <iq type='set'>
    <query xmlns='urn:xmpp:mam:2' queryid={Math.random()} />
  </iq>
)

// otherwise
const result = await client.iqCaller.request(
  xml('iq', {type: 'set'},
    xml('query', {xmlns: 'urn:xmpp:mam:2', queryid: Math.random()})
  )
)

console.log(result.toString())
```

If you think xmpp.js lacks a primitive that belongs there please let me know. For example xmpp.js doesn't support stanza streams (multiple stanzas related to each others) but that could be added if needed.

See

* https://github.com/xmppjs/xmpp.js/tree/main/packages/iq
* https://github.com/xmppjs/xmpp.js/tree/main/packages/xml

**Reply by thegodtune** [reply]
Wow, thank you very much, Sonny, I will spend time studying this response from you.

Thank you for being so helpful.

**Reply by sonnyp** [reply]
Here is some old code I dug from my archives - feel free to use it but please add my name to credits/contributors.



```js
export async function loadMessages({before, start, end, after, max, ...args}) {
  const result = await iqCaller.request(
    <iq type="set">
      <query xmlns="urn:xmpp:mam:2">
        <x xmlns="jabber:x:data" type="submit">
          <field var="FORM_TYPE" type="hidden">
            <value>urn:xmpp:mam:2</value>
          </field>
          {args.with && (
            <field var="with">
              <value>{args.with}</value>
            </field>
          )}
          {start && (
            <field var="start">
              <value>{new Date(start).toISOString()}</value>
            </field>
          )}
          {end && (
            <field var="end">
              <value>{new Date(end).toISOString()}</value>
            </field>
          )}
        </x>
        <set xmlns="http://jabber.org/protocol/rsm">
          {max && <max>{max}</max>}
          {typeof before === 'string' && <before>{before}</before>}
          {typeof after === 'string' && <after>{after}</after>}
        </set>
      </query>
    </iq>,
  );
  return result.getChild('fin', 'urn:xmpp:mam:2');
}
```

```js
xmpp.on('stanza', (stanza) => {
  if (!stanza.is('message')) return;
  if (stanza.attrs.type === 'error') return;

  // other kinds of messages
  // ...

  // mam result
  const result = stanza.getChild('result', 'urn:xmpp:mam:2');
  if (result) {
    const forwarded = result.getChild('forwarded', 'urn:xmpp:forward:0');
    if (!forwarded) return;
    const delay = forwarded.getChild('delay', 'urn:xmpp:delay');
    const message = forwarded.getChild('message');
    if (!message) return;

    const date = delay?.attrs.stamp ? new Date(delay?.attrs.stamp) : undefined;
    if (message.getChildText('body')) {
      // message is a mam message
    }
  }
});
```

**Reply by thegodtune** [reply]
Wow!
Thank you very much @sonnyp, I was away to get a grip of my health.
I will definitely give you the deserved credits, whatever we are building today is possible because of the intelligent work you already did with xmpp.js
Thank you
